### ISSUE TITLE
Segmentation fault at line 756 of ngx_http_ubus_module.c

### ISSUE THREAD

**Issue by tobiaswaldvogel** [Ansuel/nginx-ubus-module#2]
I'm using OpenWRT with nginx and luci_statistics on one of my devices and I am collecting data for quite a lot of other devices.

After adding another device the statistics overview page stopped working and I found messages in the syslog about signal 11 (SIGSEV) for nginx worker processes.
After compiling nginx with debug symbols and installing libc with debug symbols I was able to identify the origin with valgrind in line 756 of ngx_http_ubus_module.c, by the looks because in->buf->pos = 0.

After reading a little bit through the nginx API for ngx_http_read_client_request_body I figured out that this is related to client_body_buffer_size, which defaults on my device to 8192. If the request is bigger (in my case 11520)  then it is stored in a temp file rather than in buffers.

From [http://nginx.org/en/docs/dev/development_guide.html](url)
"The body can be saved in memory buffers or file buffers, if the capacity specified by the client_body_buffer_size directive is not enough to fit the entire body in memory."

As a workaround I increased client_body_buffer_size to 128k, which solved the problem for me.
Probably it is not worth implementing the file read logic, as usually UBUS requests are pretty small. But I would suggest throwing an error message if in->buf->pos == 0, saying that client_body_buffer_size should be increased.

**Comment by Ansuel** [comment]
Yep I didn't implement the file logic stuff because normally you should never use ubus for very big request... wonder how uhttpd handles that... Can you check this?

**Comment by tobiaswaldvogel** [comment]
uhttpd seem to work in a complete different way, see code snippet below. As far as I have seen uh_ubus_data_send is called for each chunk of data, so it does not have this kind of problem.

Probably there are very few users running into this problem, so as said before for me the best and simplest solution would be just throwing and error message case in->buf->pos == 0 suggesting to increase client_body_buffer_size.

That would save to the time to identify the root cause for others 

```
 912 static int uh_ubus_data_send(struct client *cl, const char *data, int len)
 913 {
 914         struct dispatch_ubus *du = &cl->dispatch.ubus;
 915 
 916         if (du->jsobj || !du->jstok)
 917                 goto error;
 918 
 919         du->post_len += len;
 920         if (du->post_len > UH_UBUS_MAX_POST_SIZE)
 921                 goto error;
 922 
 923         du->jsobj = json_tokener_parse_ex(du->jstok, data, len);
 924         return len;
 925 
 926 error:
 927         uh_ubus_single_error(cl, ERROR_PARSE);
 928         return 0;
 929 }
 930 
 931 static void uh_ubus_handle_request(struct client *cl, char *url, struct path_info *pi)
 932 {
 933         struct dispatch *d = &cl->dispatch;
 934         struct dispatch_ubus *du = &d->ubus;
 935         char *chr;
 936 
 937         du->url_path = strdup(url);
 938         if (!du->url_path) {
 939                 ops->client_error(cl, 500, "Internal Server Error", "Failed to allocate resources");
 940                 return;
 941         }
 942         chr = strchr(du->url_path, '?');
 943         if (chr)
 944                 chr[0] = '\0';
 945 
 946         du->legacy = false;
 947 
 948         switch (cl->request.method)
 949         {
 950         case UH_HTTP_MSG_GET:
 951                 uh_ubus_handle_get(cl);
 952                 break;
 953         case UH_HTTP_MSG_POST:
 954                 d->data_send = uh_ubus_data_send;
 955                 d->data_done = uh_ubus_handle_post;
 956                 d->close_fds = uh_ubus_close_fds;
 957                 d->free = uh_ubus_request_free;
 958                 du->jstok = json_tokener_new();
 959                 return;
```

**Comment by ronhombre** [comment]
I'm running into this issue running Nginx 1.26.1 (currently OpenWRT latest) as well as a custom build of Nginx 1.29.2 (with OpenSSL 3.5.4) when using QUIC. Trying to do the `client_body_buffer_size 128k;` workaround fix doesn't work even if I try 1m as value. This occurs only on `/cgi-bin/luci/admin/network/dhcp` where `/ubus/?1761325457270` causes `net::ERR_QUIC_PROTOCOL_ERROR.QUIC_PUBLIC_RESET` on Google Chrome (latest).

Nginx Error(debug logs on):
```
2025/10/25 01:04:17 [notice] 13827#0: signal 17 (SIGCHLD) received from 14889
2025/10/25 01:04:17 [alert] 13827#0: worker process 14889 exited on signal 11
```

**Comment by Ansuel** [comment]
Ok interesting I will try to reproduce!

Il Ven 24 Ott 2025, 19:07 Ron Lauren Hombre ***@***.***> ha
scritto:

**Comment by ronhombre** [comment]
> Ok interesting I will try to reproduce!
> 
> Il Ven 24 Ott 2025, 19:07 Ron Lauren Hombre ***@***.***> ha
> scritto:
> […](#)

You need a public CA cert to use QUIC on Chrome and Firefox. They enforce "No QUIC" using private self-signed certs. Also, you can check QUIC status using Dev Tools and open `Privacy and Security` tab. It may be hidden in `More tools` in the DevTools menu

**Comment by ronhombre** [comment]
To further help you, the Request Content Length is `1547` and the Response Content Length is `85381`. I have quite a lot of DHCP config so it may be the reason why it's so big. Currently, I'm trying to find workarounds using nginx conf

**Comment by Ansuel** [comment]
Thanks for the hint! I have a CA with let's encrypt so no problem for that.
But wasn't aware they deny self signed. That is very important.

Il Ven 24 Ott 2025, 19:54 Ron Lauren Hombre ***@***.***> ha
scritto:

**Comment by ronhombre** [comment]
Sorry for the fragmented replies.

Another side note, everything works fine using TLS 1.3 with HTTP/2. There seems to be something quirky about how Nginx processes QUIC fragments compared to TCP/TLS ones

**Comment by Ansuel** [comment]
Probably I'm not handling fragment correctly. Can you give me your DHCP
config? It should be ok if you want you can scramble the Mac address and
names. I just need the size eheheheh

Il Ven 24 Ott 2025, 20:01 Ron Lauren Hombre ***@***.***> ha
scritto:

**Comment by ronhombre** [comment]
I fixed it by starting uhttpd in different ports and just doing a `proxy_pass` for the `/ubus`.

Sorry, I don't think sharing my DHCP config is a good idea. It exposes my internal network layout.

I suggest you update your code to match how they did it in `uhttpd/ubus.c`, possibly use some scripts and git patches to put your nginx module code on top instead of maintaining a separate copy so you can merge from upstream easily.

I'm also very sorry if this comes as cold. I've been working on this Nginx stuff for a week now and I'm exhausted. (I compiled nginx 1.29.2 and Openssl 3.5.4 on OpenWRT and Alpine Linux to test pure PQC TLS certs, and then I had to figure out why QUIC wasn't working for my private certs... now this issue too)

**Comment by ronhombre** [comment]
I think I found a better fix using unix sockets and nginx. Avoids uhttpd altogether.

## This expects that you're not using uci.conf but your own nginx.conf in /etc/nginx/

### Create our unix socket directory
```sh
mkdir -p /var/run/nginx/
```

### Create `ubus.conf` in `http.d` and create a ubus http unix socket
You can copy the contents here from `luci.locations` file. Make sure to replace `/ubus` with `/`
```
server {
        listen unix:/var/run/nginx/ubus-http.sock;

        location / {
                ubus_interpreter;
                ubus_socket_path /var/run/ubus/ubus.sock;
                ubus_parallel_req 2;
        }
}
```

### modify the `location /ubus`. This is found in luci.locations file in /etc/nginx/. I moved mine into `http.d`
This will now use our unix socket.
```
location /ubus {
        proxy_pass http://unix:/var/run/nginx/ubus-http.sock;

        # Comment out unneeded direct connection
        #ubus_interpreter;
        #ubus_socket_path /var/run/ubus/ubus.sock;
        #ubus_parallel_req 2;
}
```

This is such a niche problem, but it may be a good idea to document this somewhere since QUIC will slowly become more mainstream
